File: apps/api/models/proctoring.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from enum import Enum


class ViolationType(str, Enum):
    TAB_SWITCH = "tab_switch"
    WINDOW_BLUR = "window_blur"
    MOUSE_IDLE = "mouse_idle"
    RIGHT_CLICK = "right_click"
    COPY_PASTE = "copy_paste"
    BROWSER_DEVTOOLS = "browser_devtools"
    MULTIPLE_MONITORS = "multiple_monitors"


class ProctoringStatus(str, Enum):
    ACTIVE = "active"
    PAUSED = "paused"
    ENDED = "ended"
    TERMINATED = "terminated"


class ProctoringViolation(BaseModel):
    violation_type: ViolationType
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    details: Optional[str] = None


class ProctoringAgreement(BaseModel):
    task_id: str
    camera_enabled: bool = False
    terms_accepted: bool = True
    accepted_at: datetime = Field(default_factory=datetime.utcnow)


class ProctoringSession(BaseModel):
    session_id: str
    user_id: str
    task_id: str
    status: ProctoringStatus = ProctoringStatus.ACTIVE
    camera_enabled: bool = False
    violations: list[ProctoringViolation] = Field(default_factory=list)
    started_at: datetime = Field(default_factory=datetime.utcnow)
    ended_at: Optional[datetime] = None


class StartProctoringRequest(BaseModel):
    task_id: str
    camera_enabled: bool = False


class StartProctoringResponse(BaseModel):
    session_id: str
    task_id: str
    status: ProctoringStatus
    camera_enabled: bool


class ReportViolationRequest(BaseModel):
    violation_type: ViolationType
    details: Optional[str] = None


class ProctoringSessionResponse(BaseModel):
    session_id: str
    user_id: str
    task_id: str
    status: ProctoringStatus
    camera_enabled: bool
    violation_count: int
    started_at: datetime
    ended_at: Optional[datetime] = None

File: apps/web/src/types/proctoring.ts
export type ViolationType =
  | 'tab_switch'
  | 'window_blur'
  | 'mouse_idle'
  | 'right_click'
  | 'copy_paste'
  | 'browser_devtools'
  | 'multiple_monitors';

export type ProctoringStatus = 'active' | 'paused' | 'ended' | 'terminated';

export interface ProctoringViolation {
  violation_type: ViolationType;
  timestamp: string;
  details?: string;
}

export interface ProctoringAgreement {
  task_id: string;
  camera_enabled: boolean;
  terms_accepted: boolean;
  accepted_at: string;
}

export interface ProctoringSession {
  session_id: string;
  user_id: string;
  task_id: string;
  status: ProctoringStatus;
  camera_enabled: boolean;
  violations: ProctoringViolation[];
  started_at: string;
  ended_at?: string;
}

export interface StartProctoringRequest {
  task_id: string;
  camera_enabled: boolean;
}

export interface StartProctoringResponse {
  session_id: string;
  task_id: string;
  status: ProctoringStatus;
  camera_enabled: boolean;
}

export interface ProctoringSessionResponse {
  session_id: string;
  user_id: string;
  task_id: string;
  status: ProctoringStatus;
  camera_enabled: boolean;
  violation_count: number;
  started_at: string;
  ended_at?: string;
}
